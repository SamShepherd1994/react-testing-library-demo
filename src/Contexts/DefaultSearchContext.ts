import { createContext } from 'react';

type SearchState = {
  defaultValue: string;
};

const defaultSearchState: SearchState = {
  defaultValue: '',
};

export const SearchContext = createContext<SearchState>(defaultSearchState);
