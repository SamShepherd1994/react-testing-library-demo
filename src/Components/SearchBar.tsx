import { useContext, useState } from 'react';
import { SearchContext } from '../Contexts/DefaultSearchContext';

type SearchBarProps = {
  onSearch: (search: string) => void;
};

export const SearchBar: React.FC<SearchBarProps> = ({
  onSearch,
}: SearchBarProps) => {
  const { defaultValue } = useContext(SearchContext);
  const [search, setSearch] = useState('');

  return (
    <div className='flex flex-row'>
      <input
        data-testid='search-input'
        className={`border-2 p-2 ${
          search.length === 0 ? 'border-4 border-red-500' : ''
        }`}
        value={search ? search : defaultValue}
        onChange={(e) => setSearch(e.target.value)}
      />
      <div className='mr-1' />
      <button
        data-testid='search-button'
        className='bg-green-400 rounded-md p-2'
        onClick={() => onSearch(search)}
      >
        Search
      </button>
    </div>
  );
};
