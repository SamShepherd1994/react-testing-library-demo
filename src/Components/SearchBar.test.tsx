/* eslint-disable testing-library/prefer-screen-queries */
import { fireEvent, render, RenderOptions } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { ReactElement } from 'react';
import { SearchContext } from '../Contexts/DefaultSearchContext';
import { SearchBar } from './SearchBar';

const Providers: React.FC = ({ children }) => {
  return (
    <SearchContext.Provider
      value={{ defaultValue: 'this is not the usual default value...' }}
    >
      {children}
    </SearchContext.Provider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: Providers, ...options });

describe('SearchBar', () => {
  it('Displays the default value from context when search value does not exist', () => {
    const { queryByTestId } = customRender(<SearchBar onSearch={() => {}} />);
    const Input = queryByTestId('search-input') as HTMLInputElement;
    if (!Input) {
      throw new Error('Could not find Search Input.');
    }
    expect(Input.value).toBe('this is not the usual default value...');
  });

  it('updates value shown to value input by user', () => {
    const { queryByTestId } = customRender(<SearchBar onSearch={() => {}} />);
    const Input = queryByTestId('search-input') as HTMLInputElement;
    if (!Input) {
      throw new Error('Could not find Search Input.');
    }

    fireEvent.change(Input, { target: { value: 'updated' } });
    expect(Input.value).toBe('updated');
  });

  it('invokes callback with correct search value when user clicks search button', () => {
    const callbackSpy = jest.fn();

    const { queryByTestId } = customRender(
      <SearchBar onSearch={callbackSpy} />
    );
    const Input = queryByTestId('search-input') as HTMLInputElement;
    if (!Input) {
      throw new Error('Could not find Search Input.');
    }
    const Button = queryByTestId('search-button') as HTMLButtonElement;
    if (!Button) {
      throw new Error('Could not find Search Button.');
    }

    fireEvent.change(Input, { target: { value: 'search change' } });
    userEvent.click(Button);

    expect(callbackSpy).toHaveBeenCalledWith('search change');
  });
});
