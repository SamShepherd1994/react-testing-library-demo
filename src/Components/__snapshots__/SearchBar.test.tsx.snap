// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar matches snapshot before and after user input 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row"
  >
    <input
      class="border-2 p-2 border-4 border-red-500"
      data-testid="search-input"
      value="this is not the usual default value..."
    />
    <div
      class="mr-1"
    />
    <button
      class="bg-green-400 rounded-md p-2"
      data-testid="search-button"
    >
      Search
    </button>
  </div>
</DocumentFragment>
`;

exports[`SearchBar matches snapshot before and after user input 2`] = `
<DocumentFragment>
  <div
    class="flex flex-row"
  >
    <input
      class="border-2 p-2 "
      data-testid="search-input"
      value="updated"
    />
    <div
      class="mr-1"
    />
    <button
      class="bg-green-400 rounded-md p-2"
      data-testid="search-button"
    >
      Search
    </button>
  </div>
</DocumentFragment>
`;
